### ISSUE TITLE
Panzer:  Link problem when Tpetra_INST_INT_INT=OFF

### ISSUE THREAD

**Issue by kddevin** [trilinos/Trilinos#4915]
@trilinos/panzer 

## Expectations
Panzer should build with Tpetra_INST_INT_INT=OFF

## Current Behavior
I get link errors due to use of Tpetra::MultiVector with global ordinals = int.
```
  "virtual thunk to Tpetra::MultiVector<int, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::describe(Teuchos::basic_FancyOStream<char, std::__1::char_traits<char> >&, Teuchos::EVerbosityLevel) const", referenced from:
      vtable for Tpetra::MultiVector<int, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > in libpanzer-dof-mgr.a(Panzer_DOFManager.cpp.o)
      construction vtable for Tpetra::MultiVector<int, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >-in-Tpetra::Vector<int, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > in libpanzer-dof-mgr.a(Panzer_Filtered_UniqueGlobalIndexer.cpp.o)
      vtable for Tpetra::MultiVector<int, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > in libpanzer-dof-mgr.a(Panzer_Filtered_UniqueGlobalIndexer.cpp.o)
  "virtual thunk to Tpetra::Vector<int, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::describe(Teuchos::basic_FancyOStream<char, std::__1::char_traits<char> >&, Teuchos::EVerbosityLevel) const", referenced from:
      vtable for Tpetra::Vector<int, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > in libpanzer-dof-mgr.a(Panzer_Filtered_UniqueGlobalIndexer.cpp.o)
```

## Motivation and Context
We'd like to build Trilinos with only one global ordinal type enabled.

## Definition of Done

Panzer builds and runs with Tpetra_INST_INT_INT=OFF.

## Possible Solution

I see Tpetra::MultiVector<int,int,int> in tUniqueGlobalIndexerUtilities.cpp; perhaps this is the source of the problem.  One could disable this test when Tpetra_INST_INT_INT=OFF.

Or perhaps the problem is due to some Epetra vs Tpetra issue?

## Steps to Reproduce

```
cmake \
-D Trilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON \
-D Tpetra_INST_INT_LONG_LONG:BOOL=ON \
-D Tpetra_INST_INT_INT:BOOL=OFF \
-D CMAKE_BUILD_TYPE:STRING="DEBUG" \
-D CMAKE_VERBOSE_MAKEFILE:BOOL=OFF \
\
-D TPL_ENABLE_MPI:BOOL=ON \
-D MPI_EXEC_MAX_NUMPROCS:STRING=11 \
\
-D TPL_ENABLE_BinUtils:BOOL=OFF \
-D TPL_ENABLE_Pthread:BOOL=OFF \
\
-D CMAKE_C_FLAGS:STRING="-Wall  -pedantic -Wno-unknown-pragmas -Wno-narrowing -Wno-inline -Wshadow -Wdeprecated-declarations -Wempty-body  -Wignored-qualifiers -Wmissing-field-initializers  -Wsign-compare  -Wtype-limits   -Wuninitialized -Winit-self -fstrict-aliasing -Wno-long-long" \
-D CMAKE_CXX_FLAGS:STRING="-Wall -pedantic -Wno-unknown-pragmas -Wno-narrowing -Wno-delete-non-virtual-dtor -Wno-inline -Wshadow -Wdeprecated-declarations -Wempty-body  -Wignored-qualifiers -Wmissing-field-initializers  -Wsign-compare  -Wtype-limits   -Wuninitialized -Winit-self -fstrict-aliasing" \
\
-D Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=ON \
-D Trilinos_ENABLE_TESTS:BOOL=ON \
-D Trilinos_ENABLE_EXAMPLES:BOOL=ON \
\
-D Trilinos_ENABLE_SHADOW_WARNINGS:BOOL=ON \
-D Trilinos_VERBOSE_CONFIGURE:BOOL=OFF \
-D Trilinos_ENABLE_Fortran:BOOL=OFF \
\
-D Tpetra_ENABLE_DEPRECATED_CODE:BOOL=OFF \
\
-D Trilinos_ENABLE_Stokhos:BOOL=ON \
-D Trilinos_ENABLE_Nox:BOOL=ON \
-D Trilinos_ENABLE_ROL:BOOL=ON \
-D Trilinos_ENABLE_MiniTensor:BOOL=OFF \
-D ROL_ENABLE_MiniTensor:BOOL=OFF \
-D Trilinos_ENABLE_Panzer:BOOL=ON \
-D Trilinos_ENABLE_PanzerAdaptersSTK:BOOL=OFF \
-D Trilinos_ENABLE_PanzerAdaptersIOSS:BOOL=OFF \
-D Trilinos_ENABLE_Thyra:BOOL=ON \
-D Trilinos_ENABLE_MueLu:BOOL=ON \
-D Trilinos_ENABLE_Anasazi:BOOL=ON \
-D Trilinos_ENABLE_Belos:BOOL=ON \
\
-D Teuchos_ENABLE_STACKTRACE:BOOL=OFF \
-D Teuchos_ENABLE_LONG_LONG_INT:BOOL=ON \
.. |& tee OUTPUT.CMAKE

make -j 8 |& tee OUTPUT.MAKE
```

## Your Environment

- **Relevant repo SHA1s:**  
- **Relevant configure flags or configure script:**  
- **Operating system and version:**  
- **Compiler and TPL versions:**  

## Related Issues

* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to 
* Part of 
* Composed of

**Comment by bartlettroscoe** [comment]
@rppawlo, we are getting a deprecated warning in the ATDM Trilinos builds (see #4913).  Since I think we need `long long int` I think this might mean we may get broken code when they drop support for more than one global ordinal type.

**Comment by rppawlo** [comment]
Ugh. I've been dreading this day. Panzer currently requires both Epetra and Tpetra. The library is mostly agnostic to the underlying linear algebra (except for things like gathers/scatters), but much the testing uses epetra underneath. If GO=int is required for epetra, this might take some time. Charon2 and Drekar also heavily rely on the Epetra stuff (Charon2 is still converting to Tpetra and Drekar will lose significant functionality if it can only use Tpetra). This might be a very fast fix or a slow slog to get it disabled. I'll look into this now...

**Comment by rppawlo** [comment]
Well - its a worst case scenario. Panzer supports both the type 1 and type 2 stacks (both are currently required to enable the panzer package):
- type 1: epetra/amesos/ifpack/ml/aztecoo 
- type 2: tpetra/amesos2/ifpack2/muelu/belos

For Type 1 - the panzer dof manger uses GO=int, for type 2 it uses the default GO from Tpetra (usually long long int). At runtime we allow users to jump between the two stacks from a single input parameter. The panzer objects are built with ETI, one for int and, if the tpetra GO is not int, then a second one for that GO type:
```
#include "PanzerDofMgr_config.hpp"
#include "Panzer_DOFManager.hpp"
#include "Panzer_DOFManager_impl.hpp"

template class panzer::DOFManager<int,int>;

#ifndef PANZER_ORDINAL64_IS_INT
template class panzer::DOFManager<int,panzer::Ordinal64>;
#endif
```

To drop to a single GO type, we will have to change all code in panzer and the applications that currently uses GO=int dof manager to the GO=panzer::Ordinal64 dof manager. In panzer, this is possible, but will most likely result in extra memory use and copying - e.g. copying GO types into  int arrays for epetra map construction. Indexing during gather/scatter will be similarly impacted, but this is manageable. This could impact the applications quite heavily:

1. Charon2: currently the code only uses the type 1 stack and has not yet been able to get the type 2 stack working. This is most likely due to some very specific boundary conditions that use Epetra specific code. If we change the GO type in the panzer dof manager they will have to fix any epetra specific code that uses the dof manager for the new GO type. @glhenni 
1.  Drekar: Drekar switches between type 1 and type 2 at runtime via input file often. This works fine for most problems. However for the Algebraic Flux Correction capability, it has been hard coded to epetra with GO=int dof manager. So dropping to a single ordinal will require much code change to using the other dof manager. We want to drop the type 1 stack support so this needed to happen anyway.  It may just have to be accelerated. @sconde @ssmabuza 
1. panzer library: all of the epetra versions of gather/scatter code (each with separate implementations for residual/Jacobian/tangent/Hessian. Then do the same for Dirichlet BC versions. So at least 16 evaluators to modify. We will have to remove the ETI instantiations from all dof-manager objects (blocked and non-blocked). There is also another issue that may arise. While the dof manager is the main sticking point, this may also impact the linear object factories and the ModelEvaluators in panzer. Both of these objects have epetra/tpetra specific versions. At quick glance, it looks to be simple changes.
1. From a panzer unit testing point of view probably 70% of panzer code uses epetra and the rest tpetra. Much is agnostic to the stack type, but not everything. We will have to do some grunt work here to fix the setup of problems the directly instantiate dof managers with GO=int.
1. mini-EM: panzer contains the mini-EM application as well. Recently the type 1 stack support was added to this so that we could do direct performance comparisons of the type 1 and type 2 stack. May need changes here as well. @cgcgcg @egphill 

So this looks like it will not be a quick single day fix and will most likely require coordinated changes to downstream apps. I'm starting it now on the panzer side and will push up a branch for applications to test out soon. I'm hoping that the majority of the changes are in panzer and that the apps aren't impacted too much.

Other interested parties: @eric-c-cyr @kddevin @mhoemmen @bathmatt 

An extra benefit of this work is that as we make changes to the type 1 stack, we could also start wrapping the objects so that the entire type 1 stack can be optional instead of required for panzer.

**Comment by jmgate** [comment]
Hey @rppawlo, sometime in the last week Charon started seeing this configure warning when trying to update our Trilinos fork:
```
-- Tpetra: Validate global ordinal setting ...
CMake Warning at Trilinos/packages/tpetra/CMakeLists.txt:1207 (MESSAGE):
  Tpetra requires only one global ordinal, but more than one are set:
  {int;long long}.  Building with more than one global ordinal type in Tpetra
  is deprecated.
```

We don't yet enable Tpetra explicitly, but it looks like it's getting pulled in via TrilinosCouplings.  Is this warning wrapped up in this issue?  Is there anything I should do about it in Charon now?

**Comment by rppawlo** [comment]
@jmgate - yes - unfortunately you will see that for now. I'm not sure if there's a way to disable the configure warning. Panzer currently requires both int and long long global ordinals to support both type 1 and type 2 stack. I'm working on a large set of changes to panzer to fix this. With travel coming up next week. I may not get them finished until the week after.

**Comment by mhoemmen** [comment]
THE GREAT PURGE HAS BEGUN

**Comment by jhux2** [comment]
> 5\. mini-EM: panzer contains the mini-EM application as well. Recently the type 1 stack support was added to this so that we could do direct performance comparisons of the type 1 and type 2 stack. May need changes here as well.

@rppawlo I assume link errors like the following are related?

<details>

```
[100%] Linking CXX executable PanzerMiniEM_BlockPrec.exe
CMakeFiles/PanzerMiniEM_BlockPrec.dir/main.cpp.o: In function `panzer::ArrayToFieldVector<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::buildFieldVector(Tpetra::Vector<int, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > const&) const':
/scratch/jhu/trilinos/Trilinos/packages/panzer/dof-mgr/src/Panzer_UniqueGlobalIndexer_Utilities_impl.hpp:632: undefined reference to `Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::Map(unsigned long, Teuchos::ArrayView<int const> const&, int, Teuchos::RCP<Teuchos::Comm<int> const> const&)'
/scratch/jhu/trilinos/Trilinos/packages/panzer/dof-mgr/src/Panzer_UniqueGlobalIndexer_Utilities_impl.hpp:633: undefined reference to `Tpetra::Vector<int, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::Vector(Teuchos::RCP<Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > const> const&, bool)'
/scratch/jhu/trilinos/Trilinos/packages/panzer/dof-mgr/src/Panzer_UniqueGlobalIndexer_Utilities_impl.hpp:635: undefined reference to `Tpetra::Import<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::Import(Teuchos::RCP<Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > const> const&, Teuchos::RCP<Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > const> const&)'
/scratch/jhu/trilinos/Trilinos/packages/panzer/dof-mgr/src/Panzer_UniqueGlobalIndexer_Utilities_impl.hpp:636: undefined reference to `Tpetra::DistObject<int, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::doImport(Tpetra::SrcDistObject const&, Tpetra::Import<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > const&, Tpetra::CombineMode, bool)'
CMakeFiles/PanzerMiniEM_BlockPrec.dir/main.cpp.o: In function `_ZNK6panzer18ArrayToFieldVectorIiiN6Kokkos6Compat23KokkosDeviceWrapperNodeINS1_6OpenMPENS1_9HostSpaceEEEE13getDataVectorIdNS1_11DynRankViewIdJS4_EEEEEN7Teuchos3RCPIN6Tpetra11MultiVectorIT_iiS6_EEEERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKSt3mapISN_T0_St4lessISN_ESaISt4pairISO_SR_EEE':
/scratch/jhu/trilinos/Trilinos/packages/panzer/dof-mgr/src/Panzer_UniqueGlobalIndexer_Utilities_impl.hpp:614: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::getNumVectors() const'
/scratch/jhu/trilinos/Trilinos/packages/panzer/dof-mgr/src/Panzer_UniqueGlobalIndexer_Utilities_impl.hpp:614: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::MultiVector(Teuchos::RCP<Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > const> const&, unsigned long, bool)'
/scratch/jhu/trilinos/Trilinos/packages/panzer/dof-mgr/src/Panzer_UniqueGlobalIndexer_Utilities_impl.hpp:616: undefined reference to `Tpetra::Import<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::Import(Teuchos::RCP<Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > const> const&, Teuchos::RCP<Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > const> const&)'
/scratch/jhu/trilinos/Trilinos/packages/panzer/dof-mgr/src/Panzer_UniqueGlobalIndexer_Utilities_impl.hpp:617: undefined reference to `Tpetra::DistObject<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::doImport(Tpetra::SrcDistObject const&, Tpetra::Import<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > const&, Tpetra::CombineMode, bool)'
CMakeFiles/PanzerMiniEM_BlockPrec.dir/main.cpp.o: In function `_ZNK6panzer18ArrayToFieldVectorIiiN6Kokkos6Compat23KokkosDeviceWrapperNodeINS1_6OpenMPENS1_9HostSpaceEEEE20getGhostedDataVectorIdNS1_11DynRankViewIdJS4_EEEEEN7Teuchos3RCPIN6Tpetra11MultiVectorIT_iiS6_EEEERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKSt3mapISN_T0_St4lessISN_ESaISt4pairISO_SR_EEE':
/scratch/jhu/trilinos/Trilinos/packages/panzer/dof-mgr/src/Panzer_UniqueGlobalIndexer_Utilities_impl.hpp:555: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::MultiVector(Teuchos::RCP<Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > const> const&, unsigned long, bool)'
/scratch/jhu/trilinos/Trilinos/packages/panzer/dof-mgr/src/Panzer_UniqueGlobalIndexer_Utilities_impl.hpp:583: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::MultiVector(Teuchos::RCP<Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > const> const&, unsigned long, bool)'
/scratch/jhu/trilinos/Trilinos/packages/panzer/dof-mgr/src/Panzer_UniqueGlobalIndexer_Utilities_impl.hpp:585: undefined reference to `Tpetra::Import<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::Import(Teuchos::RCP<Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > const> const&, Teuchos::RCP<Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > const> const&)'
/scratch/jhu/trilinos/Trilinos/packages/panzer/dof-mgr/src/Panzer_UniqueGlobalIndexer_Utilities_impl.hpp:586: undefined reference to `Tpetra::DistObject<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::doImport(Tpetra::SrcDistObject const&, Tpetra::Import<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > const&, Tpetra::CombineMode, bool)'
CMakeFiles/PanzerMiniEM_BlockPrec.dir/main.cpp.o: In function `panzer_stk::ParameterListCallbackBlocked<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::buildCoordinatesTpetra(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)':
/scratch/jhu/trilinos/Trilinos/packages/panzer/adapters-stk/src/Panzer_STK_ParameterListCallbackBlocked_impl.hpp:246: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::getNumVectors() const'
/scratch/jhu/trilinos/Trilinos/packages/panzer/adapters-stk/src/Panzer_STK_ParameterListCallbackBlocked_impl.hpp:249: undefined reference to `Tpetra::Vector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::get1dCopy(Teuchos::ArrayView<double> const&) const'
/scratch/jhu/trilinos/Trilinos/packages/panzer/adapters-stk/src/Panzer_STK_ParameterListCallbackBlocked_impl.hpp:252: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::getLocalLength() const'
/scratch/jhu/trilinos/Trilinos/packages/panzer/adapters-stk/src/Panzer_STK_ParameterListCallbackBlocked_impl.hpp:253: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::getVector(unsigned long) const'
/scratch/jhu/trilinos/Trilinos/packages/panzer/adapters-stk/src/Panzer_STK_ParameterListCallbackBlocked_impl.hpp:253: undefined reference to `Tpetra::Vector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::get1dCopy(Teuchos::ArrayView<double> const&) const'
/scratch/jhu/trilinos/Trilinos/packages/panzer/adapters-stk/src/Panzer_STK_ParameterListCallbackBlocked_impl.hpp:256: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::getLocalLength() const'
/scratch/jhu/trilinos/Trilinos/packages/panzer/adapters-stk/src/Panzer_STK_ParameterListCallbackBlocked_impl.hpp:257: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::getVector(unsigned long) const'
/scratch/jhu/trilinos/Trilinos/packages/panzer/adapters-stk/src/Panzer_STK_ParameterListCallbackBlocked_impl.hpp:257: undefined reference to `Tpetra::Vector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::get1dCopy(Teuchos::ArrayView<double> const&) const'
```

</details>

**Comment by rppawlo** [comment]
@jhux2 - exactly. For now you will have to have both GO types enabled in Tpetra for panzer and miniEM.

**Comment by rppawlo** [comment]
Finally have this close to ready. Still need to make corresponding changes in all the applications before we push this (want to test everything before pushing).

PR is in: #5228

**Comment by rppawlo** [comment]
With merge of PR #5228 today, we can now compile panzer and all downstream apps with Tpetra_INST_INT_INT=OFF. To enable this I did have to set these flags to prevent link issues in xpetra (this may have been fixed - haven't checked recently):
```
-D Xpetra_ENABLE_Epetra:BOOL=OFF \
-D MueLu_ENABLE_Epetra:BOOL=OFF \
```
@kddevin - do you want to verify and close?

**Comment by bartlettroscoe** [comment]
@rppawlo, we will need to verify that this works with EMPIRE and SPARC.  I don't know why they would need Epetra support but it looks like currently one of them does from the ATDM Trilinos configuration shown at, for example:

* https://testing.sandia.gov/cdash/viewConfigure.php?buildid=5172582

which shows:

```
-- Setting Xpetra_ENABLE_Epetra=ON since Trilinos_ENABLE_Xpetra=ON AND Trilinos_ENABLE_Epetra=ON
...
-- Setting MueLu_ENABLE_Epetra=ON since Trilinos_ENABLE_MueLu=ON AND Trilinos_ENABLE_Epetra=ON
```

We will have to test turning these off and going to a single global ordinal in the file [ATDMDevEnvSettings.cmake](https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/ATDMDevEnvSettings.cmake).

Will disabling these automatically avoid the extra `int` global ordinal instantiation?

**Comment by mhoemmen** [comment]
MueLu sets the rule is that if you disable `GlobalOrdinal=int`, you can't have Epetra support (in Xpetra or MueLu).

**Comment by bartlettroscoe** [comment]
Looks like EMPIRE has not updated Trilinos 'develop' since May 30.  Therefore, it will be hard to test a change like this since we can't test EMPIRE 'develop' against Trilinos 'develop'.

But I guess we can create a new GitHub issue for the instantiation for global ordinal `int` and close this one (since this is about Panzer, not ATDM APPs).

**Comment by rppawlo** [comment]
This change has already been tested with INST_INT_INT=OFF for empire, drekar and charon2. All three apps compile and pass everything against current Trilinos develop. Apps teams signed off on this PR before it was pushed to Trilinos. Not saying that we tested every configuration on every platform, but we did some extensive testing since it was so disruptive a change. @bathmatt has asked to freeze trilinos updates (including this work) into the empire Trilinos fork since their test suite is not clean yet. He wants a master promotion before accepting new trilinos changes.

**Comment by rppawlo** [comment]
I just tried removing the xpetra and muelu flags above and get this:
```
Processing enabled package: Xpetra (Libs)
-- Xpetra: Enabling deprecated code
-- Xpetra support for 32 bit Epetra enabled.
--    Xpetra_Epetra_NO_32BIT_GLOBAL_INDICES=OFF
CMake Error at packages/xpetra/CMakeLists.txt:146 (MESSAGE):
  If Xpetra Epetra support is enabled and Epetra 32-bit global indices are
  enabled, Xpetra requires that you enable Tpetra_INST_INT_INT
```
I'm not explicitly enabling xpetra-epetra support and enable_optional packages is off, so I'm not sure how the default gets set to on.

For panzer, certain applications require being able to switch between epetra-32bit+type1 stack and tpetra+type2 stack at runtime. I'll be working to allow users to disable epetra and type1 stack in panzer at configure time soon. But right now both stacks are required by panzer.

**Comment by bartlettroscoe** [comment]
@rppawlo said:

>  just tried removing the xpetra and muelu flags 

Specifically, what did you try?  Looking at the logic in that file it starts with:

```
IF (${PACKAGE_NAME}_ENABLE_Epetra)
  IF(NOT ${${PACKAGE_NAME}_Epetra_NO_32BIT_GLOBAL_INDICES})
    MESSAGE("-- Xpetra support for 32 bit Epetra enabled.")
    MESSAGE("--    ${PACKAGE_NAME}_Epetra_NO_32BIT_GLOBAL_INDICES=" ${${PACKAGE_NAME}_Epetra_NO_32BIT_GLOBAL_INDICES})
    
    # Tpetra and Epetra need to match GO types
    IF(${PACKAGE_NAME}_ENABLE_Tpetra)
      IF(NOT Tpetra_INST_INT_INT)
        MESSAGE(FATAL_ERROR "If Xpetra Epetra support is enabled and Epetra 32-bit global indices are enabled, Xpetra requires that you enable Tpetra_INST_INT_INT")
      ENDIF()
    ENDIF()
  ...
ENDIF()
```

So it is showing that Xpetra_ENABLE_Epetra is set to ON.

Let me give this a try using that ATDM Trilinos configuration.

**Comment by rppawlo** [comment]
> Specifically, what did you try?

just deleted the two flags from my configure script

**Comment by bartlettroscoe** [comment]
FYI: I just tested the tip of EMPIRE 'develop' + Trilinos 'develop':

```
---------------------------------------------------------------------------
| ID | Repo Dir      | Branch              | Tracking Branch  | C | M | ? |
|----|---------------|---------------------|------------------|---|---|---|
|  0 | EMPIRE (Base) | develop             | cgl/develop      |   |   |   |
|  1 | EMPIRE-PIC    | develop             | cgl/develop      |   |   |   |
|  2 | EMPIRE-Fluid  | develop             | cgl/develop      |   |   |   |
|  3 | EMPIRE-Hybrid | develop             | cgl/develop      |   |   |   |
|  4 | TPL/SPIN      | develop             | cgl/develop      |   |   |   |
|  5 | Trilinos      | tril-github-develop | trilinos/develop |   |   |   |
---------------------------------------------------------------------------
```

as of the commits (i.e. `EMPIRERepoVersion.txt`):

```
*** Base Git Repo: EMPIRE
9e8dfc7 [Wed Jun 12 18:40:02 2019 -0600] <chmoore@sandia.gov>
... comment removed ...
*** Git Repo: EMPIRE-PIC
4bec700 [Wed Jun 12 18:40:08 2019 -0600] <chmoore@sandia.gov>
... comment removed ...
*** Git Repo: EMPIRE-Fluid
fc6285a [Thu Jun 13 16:11:46 2019 -0600] <eccyr@sandia.gov>
... comment removed ...
*** Git Repo: EMPIRE-Hybrid
e30ae10 [Tue Jun 11 13:32:19 2019 -0600] <rppawlo@sandia.gov>
... comment removed ...
*** Git Repo: TPL/SPIN
35938fc [Mon Jun 3 09:12:51 2019 -0600] <wmcdoni@doppio.sandia.gov>
... comment removed ...
*** Git Repo: Trilinos
f95647b [Thu Jun 13 17:15:15 2019 -0600] <trilinos-autotester@trilinos.org>
Merge Pull Request #4683 from fryeguy52/Trilinos/spack-rhel6-env
```

and it built and passed all but three tests:

```
Fri Jun 14 07:22:20 MDT 2019

empire-tril-build-test.sh -p ../EMPIRE -b sems-rhel6-gnu-7.2.0-openmp-opt -a
  &> sems-rhel6-gnu-7.2.0-openmp-opt.out

real    9m43.840s
user    131m47.914s
sys     15m28.328s

99% tests passed, 3 tests failed out of 663
```

Therefore, I think we can test this against EMPIRE 'develop'.

**Comment by bartlettroscoe** [comment]
FYI: The changes to turn off global int are in PR #5384.  I am testing this right now with the native Trilinos test suite and then I will test with EMPIRE and SPARC.

**Comment by rppawlo** [comment]
Updated the required build flags in the comment above - added Piro_ENABLE_MueLu=OFF. This is needed until #5229 is addressed.

**Comment by bartlettroscoe** [comment]
@rppawlo said:

> Updated the required build flags in the comment above - added Piro_ENABLE_MueLu=OFF. This is needed until #5229 is addressed.

I will try this.  But will this not disable a bunch of Panzer tests that use MueLu or do the Panzer tests directly use Stratimikos with MueLu with `long long int` global ordinal?

**Comment by rppawlo** [comment]
> I will try this. But will this not disable a bunch of Panzer tests that use MueLu or do the Panzer tests directly use Stratimikos with MueLu with `long long int` global ordinal?

We keep the type1 and type2 stacks separate so we usually only need muelu when using tpetra with the single tpetra GO type. Recently four tests were added to mini-em that broke this convention for timing comparisons between the stacks. They are currently disabled, but @cgcgcg put up a PR request this weekend to reenable with correct logic (#5388 ).

**Comment by bartlettroscoe** [comment]
@rppawlo said:

> We keep the type1 and type2 stacks separate so we usually only need muelu when using tpetra with the single tpetra GO type.

So the type2 stack tests don't use Stratimikos at all?

**Comment by rppawlo** [comment]
Both stacks use stratimikos. Panzer registers the muelu preconditioners with stratimikos using the GO that is available in tpetra. Prior to dropping to one GO, panzer was registering both int and long long GO versions of muelu with stratimikos.

**Comment by bartlettroscoe** [comment]
We have verified that this is resolved.  We can close this as far as ATDM is concerned.

The rest of Trilinos has problems with turning of global int as per #5411, #5412, and #5414.  I suspect that if we turned on all primary tested packages we would see even more issues.

**Comment by bartlettroscoe** [comment]
This is long longer an issue for the ATDM Trilinos builds so removing the ATDM keyword. 

@kddevin, you might consider closing this.  As the original reporter of this issue, you are best qualified to verify that the problems are resolved.

**Comment by rppawlo** [comment]
This has been done for a while. Closing.
